File: Algorithm/leetcode/752.打开转盘锁.ts
/*
import { typescript } from 'rollup-plugin-typescript2';
 * @lc app=leetcode.cn id=752 lang=typescript
 *
 * [752] 打开转盘锁
 */

// @lc code=start
/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
var openLock = function (deadends: string[], target: string): number {

  function plusOne(s: string, j: number) {
    const sArr: string[] = s.split('');
    if (sArr[j] === '9') {
      sArr[j] = '0';
    } else {
      sArr[j] = (parseInt(sArr[j]) + 1).toString();
    }
    return sArr.join('');
  }
  function minusOne(s: string, j: number) {
    const sArr: string[] = s.split('');
    if (sArr[j] === '0') {
      sArr[j] = '1';
    } else {
      sArr[j] = (parseInt(sArr[j]) - 1).toString();
    }
    return sArr.join('');
  }


  const deads = new Set();
  for (const s in deadends) deads.add(s);
  const visited = new Set<string>();
  const queue = [];
  let step=0;
  queue.push("0000");
  visited.add("0000");

  while (!!queue.length) {
    let queueSize = queue.length;
    for (let i = 0;i < queueSize;i++){
      const cur = queue.pop();
      if (deads.has(cur)){
        continue;
      }
      if (cur === target) {
        return step;
      }
      for (let j = 0; j < 4; j++){
        const up = plusOne(cur, j);
        if (!visited.has(up)) {
          queue.push(up);
          visited.add(up)
        }
        const down = minusOne(cur, j);
        if (!visited.has(down)) {
          queue.push(down);
          visited.add(down)
        }
      }
    }
    step++;
  }
  return -1;

};

// @lc code=end

File: Algorithm/leetcode/752.打开转盘锁.java
import java.util.HashSet;
import java.util.Set;

/*
 * @lc app=leetcode.cn id=752 lang=java
 *
 * [752] 打开转盘锁
 */

// @lc code=start
class Solution {

    String plusOne(String s, int j){
      char [] ch = s.toCharArray();
      if(ch[j] == '9'){
        ch[j] = '0';
      }else{
        ch[j] += 1;
      }
      return new String(ch);
    }
    String minusOne(String s, int j){
      char [] ch = s.toCharArray();
      if(ch[j] == '0'){
        ch[j] = '9';
      }else{
        ch[j] -= 1;
      }
      return new String(ch);
    }
    public int openLock(String[] deadends, String target) {
      Set<String> deads = new HashSet<>();
      for(String s: deadends){
        deads.add(s);
      }
      // 记录已经穷举过的密码，防止走回头路
      Set<String> visited = new HashSet<>();
      Queue<String> q = new LinkdList<>();
      int step = 0;
      q.offer("0000");
      visited.add("0000");
      while(!q.isEmpty()){
        int sz = q.size();
        for(int i=0; i<sz; i++){
          String cur = q.poll();
          if(deads.contains(cur)){
            continue;
          }
          if(cur.equals(target)){
            return step;
          }
          // 将一个节点的未遍历相邻节点加入队列
          for(int j=0; i<4; j++){
            String up = plusOne(cur, j);
            String down = minusOne(cur, j);
            q.offer(up);
            q.offer(down);
          }
        }
        step++;
      }
      return -1;
    }
}
// @lc code=end

